array_api_tests/test_special_cases.py

```diff
@@ -0,0 +1,646 @@
+
+                    def cond(*inputs) -> bool:
+                        return unary_cond(abs(inputs[args_i]))
+
+                elif r_and_input.match(input_str):
+                    left_expr = expr_template.replace("{}", "x1ᵢ")
+                    right_expr = expr_template.replace("{}", "x2ᵢ")
+                    expr = f"({left_expr}) and ({right_expr})"
+
+                    def cond(i1: float, i2: float) -> bool:
+                        return unary_cond(i1) and unary_cond(i2)
+
+                elif r_or_input.match(input_str):
+                    left_expr = expr_template.replace("{}", "x1ᵢ")
+                    right_expr = expr_template.replace("{}", "x2ᵢ")
+                    expr = f"({left_expr}) or ({right_expr})"
+
+                    def cond(i1: float, i2: float) -> bool:
+                        return unary_cond(i1) or unary_cond(i2)
+
+                else:
+                    raise ValueParseError(input_str)
+
+            conds.append(cond)
+            cond_exprs.append(expr)
+
+    result_m = r_result.match(case_m.group(2))
+    if result_m is None:
+        raise ValueParseError(case_m.group(2))
+    result_str = result_m.group(1)
+    if m := r_array_element.match(result_str):
+        sign, input_ = m.groups()
+        result_expr = f"{sign}x{input_}ᵢ"
+        if input_ == "1":
+            if sign != "-":
+
+                def check_result(i1: float, i2: float, result: float) -> bool:
+                    eq = make_eq(i1)
+                    return eq(result)
+
+            else:
+
+                def check_result(i1: float, i2: float, result: float) -> bool:
+                    eq = make_eq(-i1)
+                    return eq(result)
+
+        else:
+            if sign != "-":
+
+                def check_result(i1: float, i2: float, result: float) -> bool:
+                    eq = make_eq(i2)
+                    return eq(result)
+
+            else:
+
+                def check_result(i1: float, i2: float, result: float) -> bool:
+                    eq = make_eq(-i2)
+                    return eq(result)
+
+    else:
+        _check_result, result_expr = parse_result(result_m.group(1))
+
+        def check_result(i1: float, i2: float, result: float) -> bool:
+            return _check_result(result)
+
+    expr = " and ".join(cond_exprs) + " -> " + result_expr
+
+    def cond(i1: float, i2: float) -> bool:
+        return all(c(i1, i2) for c in conds)
+
+    return BinaryCase(expr, cond, check_result)
+
+
+r_redundant_case = re.compile("result.+determined by the rule already stated above")
+
+
+def parse_binary_docstring(docstring: str) -> List[BinaryCase]:
+    match = r_special_cases.search(docstring)
+    if match is None:
+        return []
+    lines = match.group(1).split("\n")[:-1]
+    cases = []
+    for line in lines:
+        if m := r_case.match(line):
+            case_str = m.group(1)
+        else:
+            warn(f"line not machine-readable: '{line}'")
```

Comment:
We really want this to be an error, not a warning, I think.

Reply:
And even if we didn't want this to be an error, I would avoid using `warnings`. pytest is just going to list warnings at the end, but modules like NumPy already emit dozens of warnings when running the tests, which can be ignored, and any legitimate warning would be completely lost.

Reply:
I'd rather not raise an error here as it's not critical (unlike the assertions I make in `stubs.py`), so would unnecessarily halt use of the test suite for the myriad of reasons a user could have an out-of-sync `array-api` submodule.

I agree warnings are pretty awkward in practice, but I still think they're the only appropriate choice we have. I was actually using this in `setup.cfg` when developing all the parsing code paths:
```python
[tool:pytest]
filterwarnings =
    ignore::ImportWarning
    ignore::RuntimeWarning
    ignore::DeprecationWarning
```